### PR TITLE
fix: implement all 20 audit findings (2 High, 5 Medium, 13 Low)

### PR DESCRIPTION
**Pull request by kaiohenricunha**
## Summary

Full implementation of the codebase audit. All 20 findings addressed via strict TDD (red → green → commit per finding). Spec: `docs/audit-findings-spec.md`.

### High severity
- **SEC-01** — SSRF via legacy security mode: block `169.254.0.0/16` link-local in all modes, deprecate legacy mode, block redirects unconditionally
- **SEC-02** — No custom TLS for HTTPS scrape targets: new `pkg/aggregator/tls.go` with `BuildTLSConfig()` reading `METRICS_SCRAPE_TLS_*` env vars; wired into `NewAggregator`

### Medium severity
- **BUG-01 + ENH-03** — Histogram sum/count not atomically consistent + O(N) observe: unified `sync.Mutex` for all counts and sum; binary search via `sort.SearchFloat64s`
- **BUG-03** — `sampleLine` regex rejects valid metrics with `}` in label values: quoted-string-aware pattern
- **ENH-01** — No cache hit/miss metrics: `metrics_aggregator_cache_hits_total` / `cache_misses_total`
- **ENH-02** — No build version info metric: `metrics_aggregator_build_info{version,commit}` + Dockerfile ldflags + CI build-args
- **SEC-03** — E2E binary downloads lack checksum verification: `sha256sum -c` in `lib.sh`
- **SEC-04** — Dockerfile uses mutable image tags: both `FROM` lines pinned to `@sha256:` digest

### Low severity
- **BUG-02** — Oversized body incorrectly increments `scrape_errors_total`: `failed = false` on size-limit discard
- **BUG-04** — `X-Request-Id` not length/char validated: truncate to 128, strip non-printable, fallback on empty
- **SEC-05** — JSON parse error log reflects config structure: neutral message, no `error` or `input_length` fields
- **SEC-06** — Port not validated: `validatePort()` enforces 1–65535
- **SEC-07** — No bind-address restriction: `METRICS_BIND_ADDRESS` env var
- **ENH-04** — No stale-cache-on-error: `METRICS_CACHE_SERVE_STALE_ON_ERROR` with `lastGoodValue`
- **ENH-05** — HTTPS test gap: `httptest.NewTLSServer` coverage
- **ENH-06** — Env parser edge cases: negative duration, zero/negative int tests
- **ENH-07** — Context cancellation skips caching: propagate `ctx.Err()` without storing result
- **ENH-08** — `promtool` lint filter too broad: tightened to exact pattern
- **ENH-09** — `prom/prometheus:latest` in docker-compose: pinned to `v3.3.0`

## Test plan

- [x] `go test -race -count=1 ./...` — 119 tests, 0 failures, 0 races
- [x] `staticcheck ./...` — clean
- [x] `make smoke` — Docker Compose stack: all containers healthy, `origin_container` labels present, self-instrumentation metrics present, both Prometheus instances healthy

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: pkg/aggregator/tls.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+		cert, err := tls.LoadX509KeyPair(certFile, keyFile)
+		if err != nil {
+			return nil, fmt.Errorf("loading TLS cert/key pair: %w", err)
+		}
+		cfg.Certificates = []tls.Certificate{cert}
+	}
+
+	if skipVerify == "true" {
+		cfg.InsecureSkipVerify = true // #nosec G402 -- controlled by explicit operator opt-in
+	}
```
[the remark this answers is not in the document]
Comment (kaiohenricunha) [reply pkg/aggregator/tls.go:50]: Fixed in commit a9a226a: added `log.Warn().Msg("METRICS_SCRAPE_TLS_INSECURE_SKIP_VERIFY=true: TLS certificate verification disabled")` immediately after setting InsecureSkipVerify. Also added `TestBuildTLSConfig_InsecureSkipVerify_LogsWarning` to assert the warning is emitted.

File: main.go [hunk 2]
```diff
@@ -402,7 +482,11 @@ func main() {
 	if port == "" {
 		port = aggregator.DefaultAggregatorPort
 	}
-	addr := ":" + port
+	if _, err := validatePort(port); err != nil {
+		log.Fatal().Err(err).Msg("invalid METRICS_AGGREGATOR_PORT")
+	}
+	bindAddr := os.Getenv("METRICS_BIND_ADDRESS")
+	addr := bindAddr + ":" + port
```
[the remark this answers is not in the document]
Comment (kaiohenricunha) [reply main.go]: Fixed in commit a9a226a: both reads now use `strings.TrimSpace(os.Getenv(...))`. Added `TestPortValidation_LeadingTrailingSpaces` to cover the `" 9090 "` → `"9090"` case.

File: Dockerfile [hunk 3]
```diff
@@ -1,15 +1,19 @@
...
+FROM golang:${GO_VERSION}-alpine@sha256:2389ebfa5b7f43eeafbd6be0c3700cc46690ef842ad962f6c5bd6be49ed82039 AS builder
 WORKDIR /app
 RUN apk add --no-cache git
 COPY go.mod go.sum ./
 RUN go mod download
 COPY . .
-RUN CGO_ENABLED=0 go build -o metrics-aggregator .
+ARG VERSION=dev
+ARG COMMIT=unknown
+RUN CGO_ENABLED=0 go build -ldflags="-X main.version=${VERSION} -X main.commitSHA=${COMMIT}" -o metrics-aggregator .
```
[the remark this answers is not in the document]
Comment (kaiohenricunha) [reply Dockerfile:12]: Fixed in commit a9a226a: removed the duplicate outer `ARG VERSION=dev` / `ARG COMMIT=unknown` declarations (before `FROM`) since they aren't used by the `FROM` instruction and are re-declared inside the builder stage. Added a comment clarifying the digest pin is authoritative over the `GO_VERSION` arg.

File: pkg/aggregator/aggregator_test.go [hunk 4]
```diff
@@ -933,3 +934,168 @@ func TestAggregator_NoSpansWhenNoTracer(t *testing.T) {
...
+	t.Setenv("METRICS_SCRAPE_TLS_KEY", "")
+	t.Setenv("METRICS_SCRAPE_TLS_INSECURE_SKIP_VERIFY", "")
+
+	cfg, err := BuildTLSConfig()
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	if cfg == nil {
+		t.Fatal("expected non-nil TLS config")
+	}
```
[the remark this answers is not in the document]
Comment (kaiohenricunha) [reply pkg/aggregator/aggregator_test.go]: Fixed in commit a9a226a: renamed to `TestBuildTLSConfig_DefaultsTLS12`, which accurately describes what is asserted (non-nil config with `MinVersion=TLS12`).

File: test/e2e/lib.sh [hunk 5]
```diff
@@ -162,8 +162,22 @@ ensure_promtool() {
...
-  curl -sL "$url" | tar xz --strip-components=1 -C /tmp "prometheus-2.53.0.${os}-${arch}/promtool"
+  local base_url="https://github.com/prometheus/prometheus/releases/download/v2.53.0"
+  local url="${base_url}/${tarball}"
+  local checksum_url="${base_url}/sha256sums.txt"
+  local tmp_tar; tmp_tar="$(mktemp /tmp/promtool-XXXXXX.tar.gz)"
+  local tmp_sums; tmp_sums="$(mktemp /tmp/promtool-sums-XXXXXX.txt)"
+  curl -sL "$url" -o "$tmp_tar"
+  curl -sL "$checksum_url" -o "$tmp_sums"
+  # Verify checksum — extract the line for this tarball and check
+  if ! grep "${tarball}" "$tmp_sums" | sed "s|${tarball}|${tmp_tar}|" | sha256sum -c --status; then
```
[the remark this answers is not in the document]
Comment (kaiohenricunha) [reply test/e2e/lib.sh]: Fixed in commit a9a226a: replaced the `grep | sed | sha256sum -c` pipeline with explicit `awk`-based fixed-string matching. If the tarball entry is absent from sha256sums.txt the function now prints a clear error and returns 1 before calling sha256sum.

File: test/e2e/lib.sh [hunk 6]
```diff
@@ -162,8 +162,22 @@ ensure_promtool() {
...
-  local url="https://github.com/prometheus/prometheus/releases/download/v2.53.0/${tarball}"
-  curl -sL "$url" | tar xz --strip-components=1 -C /tmp "prometheus-2.53.0.${os}-${arch}/promtool"
+  local base_url="https://github.com/prometheus/prometheus/releases/download/v2.53.0"
+  local url="${base_url}/${tarball}"
+  local checksum_url="${base_url}/sha256sums.txt"
+  local tmp_tar; tmp_tar="$(mktemp /tmp/promtool-XXXXXX.tar.gz)"
+  local tmp_sums; tmp_sums="$(mktemp /tmp/promtool-sums-XXXXXX.txt)"
+  curl -sL "$url" -o "$tmp_tar"
+  curl -sL "$checksum_url" -o "$tmp_sums"
+  # Verify checksum — extract the line for this tarball and check
```
[the remark this answers is not in the document]
Comment (kaiohenricunha) [reply test/e2e/lib.sh]: Fixed in commit a9a226a: all four `curl -sL` calls (two in `ensure_promtool`, two in `ensure_istioctl`) changed to `curl -fsSL` so HTTP 4xx/5xx responses are treated as errors.

File: .claude/settings.json [hunk 7]
```diff
@@ -13,6 +13,15 @@
             "command": "python3 -c \"import sys, json, subprocess; d=json.load(sys.stdin); fp=(d.get('tool_input') or {}).get('file_path') or ''; fp.endswith('.go') and subprocess.run(['go', 'vet', './...'])\""
           }
         ]
+      },
+      {
+        "matcher": "Edit|Write",
+        "hooks": [
+          {
+            "type": "command",
+            "command": "bash /mnt/storage/Projects/kaiohenricunha/metrics-aggregator/.claude/hooks/container-rebuild-reminder.sh"
```
[the remark this answers is not in the document]
Comment (kaiohenricunha) [reply .claude/settings.json]: Fixed in commit a9a226a: changed the absolute path to the relative `bash .claude/hooks/container-rebuild-reminder.sh` so the hook works from any clone location.

File: pkg/aggregator/aggregator.go [hunk 8]
```diff
@@ -124,11 +123,19 @@ func NewAggregator(endpointsConfig string) (*Aggregator, error) {
...
+	if err != nil {
+		return nil, fmt.Errorf("TLS configuration error: %w", err)
+	}
+	transport := &http.Transport{
+		TLSClientConfig: tlsCfg,
+	}
+	client := &http.Client{
+		Timeout:   5 * time.Second,
+		Transport: transport,
+		CheckRedirect: func(_ *http.Request, _ []*http.Request) error {
```
[the remark this answers is not in the document]
Comment (kaiohenricunha) [reply pkg/aggregator/aggregator.go]: Fixed in commit a9a226a: replaced the zero-value `&http.Transport{}` with `http.DefaultTransport.(*http.Transport).Clone()` and then set `TLSClientConfig` on the clone, preserving all DefaultTransport defaults (proxy, connection pooling, keepalives, dial timeouts, HTTP/2).

File: pkg/aggregator/aggregator.go [hunk 9]
```diff
@@ -433,6 +446,30 @@ func sanitizeURLForLog(rawURL string) string {
...
+		hostname = h
+	}
+	linkLocal := &net.IPNet{
+		IP:   net.ParseIP("169.254.0.0"),
+		Mask: net.CIDRMask(16, 32),
+	}
+	addrs, err := net.DefaultResolver.LookupIPAddr(context.Background(), hostname)
+	if err != nil {
+		// DNS failure — let the HTTP client handle it
+		return nil
```
[the remark this answers is not in the document]
Comment (kaiohenricunha) [reply pkg/aggregator/aggregator.go]: Fixed in commit a9a226a: added a 3-second context deadline (`context.WithTimeout(context.Background(), 3*time.Second)`) to the `LookupIPAddr` call in `rejectLinkLocal`.

File: main_test.go [hunk 10]
```diff
@@ -897,6 +897,16 @@ func TestPortValidation_Valid(t *testing.T) {
...
 }
 
+func TestPortValidation_LeadingTrailingSpaces(t *testing.T) {
+	p, err := validatePort(strings.TrimSpace(" 9090 "))
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	if p != "9090" {
+		t.Fatalf("expected '9090', got %q", p)
+	}
```
[the remark this answers is not in the document]
Comment (kaiohenricunha) [reply main_test.go:912]: Fixed in commit 451e538: the test now first passes the raw `" 9090 "` to `validatePort` and asserts it is **rejected** (proving trimming is necessary), then calls `validatePort(strings.TrimSpace(" 9090 "))` and asserts it is accepted.

File: pkg/aggregator/aggregator.go [hunk 11]
```diff
@@ -127,9 +127,8 @@ func NewAggregator(endpointsConfig string) (*Aggregator, error) {
 	if err != nil {
 		return nil, fmt.Errorf("TLS configuration error: %w", err)
 	}
-	transport := &http.Transport{
-		TLSClientConfig: tlsCfg,
-	}
+	transport := http.DefaultTransport.(*http.Transport).Clone()
```
[the remark this answers is not in the document]
Comment (kaiohenricunha) [reply pkg/aggregator/aggregator.go]: Fixed in commit 451e538: guarded the type assertion with `ok` check. If `http.DefaultTransport` is not a `*http.Transport`, `NewAggregator` returns an error (consistent with the TLS error handling two lines above) rather than panicking. The suggested fallback of `&http.Transport{}` was avoided since it would silently drop all transport defaults.
